here-naksha-lib-model/src/commonMain/kotlin/naksha/model/request/PropertyFilter.kt

```diff
@@ -0,0 +1,89 @@
+package naksha.model.request
+
+import naksha.base.AnyObject
+import naksha.base.Platform
+import naksha.jbon.JbFeatureDecoder
+import naksha.model.request.query.*
+
+class PropertyFilter(val req: ReadFeatures) : ResultFilter {
+
+    companion object {
+        const val PROPERTIES = "properties"
+    }
+
+    override fun call(a1: ResultTuple): ResultTuple? {
+        val pSearch = req.query.properties ?: return a1
+        if (a1.tuple == null) return null
+        if (a1.tuple!!.feature == null) return null
+        val decoder = JbFeatureDecoder()
+        decoder.mapBytes(a1.tuple!!.feature!!)
+        if (resolvePropsQuery(pSearch, decoder)) return a1
+        return null
+    }
+
+    private fun resolvePropsQuery(pQuery: IPropertyQuery?, decoder: JbFeatureDecoder) : Boolean {
+        when (pQuery) {
+            null -> return true
+            is PAnd -> {
+                pQuery.forEach {
+                    if (!resolvePropsQuery(it, decoder)) return false
+                }
+                return true
+            }
+            is POr -> {
+                pQuery.forEach {
+                    if (resolvePropsQuery(it, decoder)) return true
+                }
+                return false
+            }
+            is PNot -> return !resolvePropsQuery(pQuery.query, decoder)
+            is PQuery -> {
+                val propFromFeature = decoder.get(PROPERTIES,*pQuery.property.path.filterNotNull().toTypedArray())
+                val op = pQuery.op
+                return resolveEachOp(op,propFromFeature,pQuery.value)
+            }
+        }
+        return false
+    }
+
+    private fun resolveEachOp(op: AnyOp, featureProperty: Any?, queryProperty: Any?) : Boolean {
+        when (op) {
+            AnyOp.EXISTS -> return featureProperty != Platform.UNDEFINED
+            AnyOp.IS_NULL -> return featureProperty == null
+            AnyOp.IS_NOT_NULL -> return featureProperty != null
+            AnyOp.IS_TRUE -> return featureProperty == true
+            AnyOp.IS_FALSE -> return featureProperty == false
+            AnyOp.IS_ANY_OF -> {
+                if (queryProperty is Array<*>) return queryProperty.contains(featureProperty)
+                if (queryProperty is List<*>) return queryProperty.contains(featureProperty)
+                return false
+            }
+            AnyOp.CONTAINS -> return resolveContains(featureProperty, queryProperty)
+            StringOp.EQUALS -> return (featureProperty is String) && (queryProperty is String) && (featureProperty.toString() == queryProperty.toString())
+            StringOp.STARTS_WITH -> return (featureProperty is String) && (queryProperty is String) && (featureProperty.startsWith(queryProperty.toString()))
+            DoubleOp.EQ -> return (featureProperty is Number) && (queryProperty is Number) && (featureProperty.toDouble() == queryProperty.toDouble())
```

Comment:
I would suggest to throw exception when feature or query properties don't match operator type, to early notify developer that the code is wrong

Reply:
A filter must not throw an exception, it would break the whole request, the goal of the filter is exactly to remove features that do not match, and a wrong value type is simply not matching.

So, I rather agree to the code, when the types do not match, reject the feature (filter out).

Another improvement would be, to let the user of the filter decide what to do in error case by providing a callback-handler to the constructor of the filter, that can act when a value is not of expected type. In other words, make a default compare is not possible method (as lambda, e.g. `var compareNotMatching: Fn2<Boolean,Any?,Any?>? = null` which stands for `fun(expect: Any?, found: Any?)`). The default, when the method reference is _null_, should be _false_, so `if (compareNotMatching(a,b) ?: false)`. This allows to override it as optional constructor parameter or at runtime. For testing, you could make some exception throwing methods, but in production, you do not want that, but maybe you want to log the happening?

Reply:
As agreed per private discussion, throwing exceptions and leave a backlog item to implement the handler